lib/Transforms/SimplifyParsers.cpp

```diff
@@ -0,0 +1,168 @@
+#include <llvm/Support/ErrorHandling.h>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "p4mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "p4hir-simplify-parsers"
+
+using namespace mlir;
+
+namespace P4::P4MLIR {
+#define GEN_PASS_DEF_SIMPLIFYPARSERS
+#include "p4mlir/Transforms/Passes.cpp.inc"
+
+namespace {
+struct SimplifyParsers : public impl::SimplifyParsersBase<SimplifyParsers> {
+    void runOnOperation() override;
+
+ private:
+    /// Finds all states reachable from the 'start' state.
+    llvm::DenseSet<P4HIR::ParserStateOp> findReachableStates(P4HIR::ParserOp &parser,
```

Comment:
P4C mainline has a way to build "parser call graph". I'm wondering if it would be possible to build something similar here? I have not looked if MLIR callgraph functionality is generic enough for this.